include/albatross/models/sparse_gp.h

```diff
@@ -0,0 +1,243 @@
+      K_ff_diag += targets.covariance.diagonal();
+    }
+    Eigen::MatrixXd K_fu = this->covariance_function_(features, u);
+    Eigen::MatrixXd K_uu = this->covariance_function_(u);
+
+    auto K_uu_llt = K_uu.llt();
+    // P is such that:
+    //     Q_ff = K_fu K_uu^-1 K_uf
+    //          = K_fu L^-T L^-1 K_uf
+    //          = P^T P
+    Eigen::MatrixXd P = K_uu_llt.matrixL().solve(K_fu.transpose());
+
+    // Efficiently compute the diagonal diag[Q_ff].
+    Eigen::VectorXd Q_ff_diag = P.colwise().squaredNorm();
+
+    Eigen::VectorXd A = K_ff_diag - Q_ff_diag;
+
+    if (A.minCoeff() < 1e-6) {
+      // It's possible that the inducing points will perfectly describe
+      // some of the data, in which case we need to add a bit of extra
+      // noise to make sure lambda is invertible.
+      A += 1e-6 * Eigen::VectorXd::Ones(n);
+    }
+    /*
+     *
+     * The end goal here is to produce a vector, v, and matrix, C, such that
+     * for a prediction, f*, we can do,
+     *
+     *     [f*|f=y] ~ N(K_*u * v , K_** - K_*u * C^-1 * K_u*)
+     *
+     *  and it would match the desired prediction described above,
+     *
+     *     [f*|f=y] ~ N(K_*u S K_uf^-1 A^-1 y, K_** − Q_** + K_*u S K_u*)
+     *
+     *  we can find v easily,
+     *
+     *     v = S K_uf^-1 A^-1 y
+     *
+     *  and to get C we need to do some algebra,
+     *
+     *     K_** - K_*u * C * K_u* = K_** - Q_** + K_*u S K_u*
+     *                            = K_** - K_*u (K_uu^-1 - S) K_u*
+     *  which leads to:
+     *     C^-1 = (K_uu^-1 - S)^-1
+     *                                                  (Expansion of S)
+     *          = (K_uu^-1 - (K_uu + K_uf A^-1 K_fu)^-1)^-1
+     *                                          (Matrix Inversion Lemma)
+     *          = (K_uu^-1 K_uf (A + K_fu K_uu^-1 K_uf)^-1 K_fu K_uu-1)
+     *                                   (LL^T = K_uu and P = L^-1 K_uf)
+     *          = L^-T P (A + P^T P)^-1 P^T L^-1
+     *                                        (Searle Set of Identities)
+     *          = L^-T C A^-1 P^T (I + P A^-1 P)^-1 L^-1
+     *                           (B = (I + P A^-1 P) and R = A^-1/2 P^T)
+     *          = L^-T R^T R B^-1 L^-1
+     *
+     *  reusing some of the precomputed values there leads to:
+     *
+     *     v = L^-T B^-1 P * A^-1 y
+     */
+    Eigen::VectorXd A_sqrt = A.array().sqrt();
+    Eigen::MatrixXd RtR = A_sqrt.asDiagonal().inverse() * P.transpose();
+    RtR = RtR.transpose() * RtR;
+    Eigen::MatrixXd B = Eigen::MatrixXd::Identity(m, m) + RtR;
+
+    auto B_ldlt = B.ldlt();
+
+    Eigen::VectorXd v = P * (A.asDiagonal().inverse() * targets.mean);
+    v = B_ldlt.solve(v);
+    v = K_uu_llt.matrixL().transpose().solve(v);
+
+    Eigen::MatrixXd L_uu_inv =
+        K_uu_llt.matrixL().solve(Eigen::MatrixXd::Identity(m, m));
+    Eigen::MatrixXd RtRBiLi = RtR * B_ldlt.solve(L_uu_inv);
+    Eigen::MatrixXd C =
+        (K_uu_llt.matrixL().transpose().solve(RtRBiLi)).inverse();
```

Comment:
I think we can save this `inverse` by recognizing that `C=(L^-T R^T R B^-1 L^-1)^{-1}`=L B (R^T R)^{-1}  L^T`. Considering `R` is made up of `P`, which is made up of `K_uf`, I don't think `R` will be invertible, so we can't get rid of the inner `ldlt`. Nonetheless, an `ldlt` is a little cheaper than an inversion...

Reply:
Otherwise, this comes back to the issue I raised in the other PR whereby we could simply transport `C^{-1}` rather than `C.ldlt()` (which in this case requires an inversion and an `ldlt` decomposition as opposed to nothing).

Reply:
Yeah, good catch.  For now I removed the inverse following your suggestion.